src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -260,16 +276,74 @@ int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t
                     gavl_left_rotation(cur_node, rnode);
             }
 
-            if (!STACK_EMPTY(node_stack)) {
-                STACK_POP(node_stack, cur_node);
-                goto stack_recovery;
+            if (!TREE_STACK_IS_EMPTY(tree_ptr)) {
+                TREE_STACK_POP(tree_ptr, cur_node);
+                continue;
             } else {
                 break;
             }
         } while (1);
 
-        while (gavl_tree_iptr->root->parent != NULL)
-            gavl_tree_iptr->root = gavl_tree_iptr->root->parent;
+        while (tree_ptr->root && tree_ptr->root->parent != NULL)
+            tree_ptr->root = tree_ptr->root->parent;
+    }
+
+    return;
+}
+
+/*
+ * MPL_gavl_tree_insert
+ * Description: insert a node with key (addr, len) into gavl tree. If new node is duplicate,
+ *              we should not insert it and need to free the node and return.
+ * Parameters:
+ * gavl_tree        - (IN) gavl tree object
+ * addr             - (IN) input buffer starting addr
+ * len              - (IN) input buffer length
+ * val              - (IN) buffer object
+ */
+int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len,
+                         const void *val)
+{
+    int mpl_err = MPL_SUCCESS;
+    gavl_tree_node_s *node_ptr;
+    gavl_tree_s *tree_ptr = (gavl_tree_s *) gavl_tree;
+
+    node_ptr = (gavl_tree_node_s *) MPL_malloc(sizeof(gavl_tree_node_s), MPL_MEM_OTHER);
+    if (node_ptr == NULL) {
+        mpl_err = MPL_ERR_SHM_NOMEM;
+        goto fn_fail;
+    }
+
+    GAVL_TREE_NODE_INIT(node_ptr, addr, len, val);
+
+    if (tree_ptr->root == NULL) {
+        tree_ptr->root = node_ptr;
+    } else {
+        gavl_tree_node_s *pnode;
+        int direction;
+
+        /* search the node which will become the parent of new node */
+        pnode = gavl_tree_search_internal(tree_ptr, (uintptr_t) node_ptr->addr, node_ptr->len,
+                                          SUBSET_SEARCH);
+
+        /* find which side the new node should be inserted */
+        direction = gavl_subset_cmp_func(pnode, (uintptr_t) node_ptr->addr, node_ptr->len);
+        if (direction == BUFFER_MATCH) {
+            /* new node is duplicate, we need to delete new node and exit */
+            tree_ptr->gavl_free_fn((void *) node_ptr->val);
+            MPL_free(node_ptr);
+            goto fn_exit;
+        }
+
+        /* insert new node into pnode */
+        if (direction == SEARCH_LEFT)
+            pnode->left = node_ptr;
+        else
+            pnode->right = node_ptr;
+        node_ptr->parent = pnode;
+
+        /* rebalance the tree */
+        gavl_tree_rebalance(tree_ptr);
```

Comment:
The `rebalance the tree ` comment seems meaningless. It is better to explain why you need rebalance the tree after insertion, same for the `delete` routine.